### PR TITLE
Implement tx size measurements in TxSeq finger tree

### LINKED ISSUE #686: Mempool: Utilize tx size in TxSeq FingerTree
From @dcoutts in relation to PR #678:

> Something we could consider for the future here: we could also augment the TxSeq finger tree with the size in bytes of the txs, and that would allow the block creation code to select the first N transactions that are less than a given byte limit (ie block size) and do so in log time rather than linear time. This means shifting work: doing less when creating blocks, and a little more elsewhere to maintain the extra finger tree meaure monoid. File a ticket so we remember to consider it later.

### PR DESCRIPTION
**Pull request by intricate**
Closes #686

### CODE REVIEW

File: ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs [hunk 1]
```diff
@@ -831,9 +834,38 @@ prop_TxSeq_lookupByTicketNo_sound smalls small =
...
         foldl' (TxSeq.:>) TxSeq.Empty $ map mkTicket haystack
 
-    mkTicket x = TxTicket x (mkTicketNo x)
+    mkTicket x = TxTicket x (mkTicketNo x) 0
     mkTicketNo = TicketNo . toEnum
 
+-- | Test that the 'fst' of the result of 'splitAfterTxSize' only contains
+-- 'TxTicket's whose summed up transaction sizes are less than or equal to
+-- that of the 'TxSizeInBytes' which the 'TxSeq' was split on.
+prop_TxSeq_splitAfterTxSize :: [Positive TxSizeInBytes] -> Property
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs]: Alternatively or *additionally*, write a `splitAfterTxSizeSpec`, which is just a simple, naive implementation (e.g., https://github.com/input-output-hk/ouroboros-network/pull/1342#discussion_r357014860), and check whether it produces the same outputs as `splitAfterTxSize` for the same inputs.

**Comment by mrBliss** [comment]
> _n.b. This isn't yet hooked up with the block forging functionality. This will be done in another PR._

I pushed two commits to this branch that hook it up. (Feel free to modify them and to rebase this branch).

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs [hunk 2]
```diff
@@ -260,6 +260,7 @@ mkNodeArgs registry cfg initState tracers btime chainDB = NodeArgs
     , blockFetchSize      = nodeBlockFetchSize
     , blockMatchesHeader  = nodeBlockMatchesHeader
     , maxUnackTxs         = 100 -- TODO
+    , maxBlockSize        = 2000000 -- TODO
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs]: @dcoutts Should this come from the genesis config too?

Also, how is this defined? I'm assuming it is the size of the block *body*, not the size of the whole block, including the header.

Comment (intricate) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node.hs]: Perhaps we should instead call this `maxBlockBodySize`, then?

Also, this value probably wouldn't directly translate to the transactions we can select from the mempool by summed byte size. The block `Body` would consist of other extra bytes from the CBOR encodings of all the payloads (`TxPayload`, `Delegation.Payload`, `Update.Payload`).

Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node.hs]: Name is fine for me. We could indeed account for these extra bytes because of CBOR tags and `listLen` bytes (I believe the number of bytes is static). As you suggested on Slack, doing that in `cardano-ledger` is less fragile. @dcoutts What do you think?

File: ouroboros-consensus/src/Ouroboros/Consensus/Mempool/TxSeq.hs [hunk 3]
```diff
@@ -160,11 +180,56 @@ splitAfterTicketNo (TxSeq txs) n =
...
+    mapTuple f (x, y) = (f x, f y)
+
+    go :: TxSizeInBytes
+       -> [TxTicket tx]
+       -> [TxTicket tx]
+       -> ([TxTicket tx], [TxTicket tx])
+    go _ accTickets [] = (reverse accTickets, [])
+    go accByteSize accTickets (t:ts)
+      | accByteSize + txTicketTxSizeInBytes t <= n =
+        go (accByteSize + txTicketTxSizeInBytes t) (t:accTickets) ts
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Mempool/TxSeq.hs]: Suggestion:
```haskell
      | let accByteSize' = accByteSize + txTicketTxSizeInBytes t
      , accByteSize' <= n
      = go accByteSize' (t:accTickets) ts
```

File: ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs [hunk 4]
```diff
@@ -295,9 +296,10 @@ data LeaderResult blk =
 forkBlockProduction
     :: forall m peer blk.
        (IOLike m, ProtocolLedgerView blk, ApplyTx blk)
-    => InternalState m peer blk
+    => Word32  -- ^ Max block size
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs]:
```suggestion
    => Word32  -- ^ Max block body size
```

File: ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs [hunk 5]
```diff
@@ -831,9 +832,58 @@ prop_TxSeq_lookupByTicketNo_sound smalls small =
...
+    splitAndCheck :: TxSizeInBytes -> Bool
+    splitAndCheck txsz =
+      let (before, _after) = splitAfterTxSize txseq txsz
+      in  checkTxSizeSumLessThanOrEqual (txTickets before) txsz
+
+    txSizeSum :: [TxTicket tx] -> TxSizeInBytes
+    txSizeSum = sum . map txTicketTxSizeInBytes
+
+    checkTxSizeSumLessThanOrEqual :: [TxTicket tx] -> TxSizeInBytes -> Bool
+    checkTxSizeSumLessThanOrEqual txs txsz = txSizeSum txs <= txsz
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs]: I'm not sure whether this deserves a separate function.

File: ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs [hunk 6]
```diff
@@ -831,9 +832,58 @@ prop_TxSeq_lookupByTicketNo_sound smalls small =
...
+
+-- | Test that the results of 'splitAfterTxSizeSpec', a specification of
+-- 'splitAfterTxSize', match those of the real 'splitAfterTxSize'
+-- implementation.
+prop_TxSeq_splitAfterTxSizeSpec :: [Small TxSizeInBytes] -> Property
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs]: Why `Small` instead of `Positive`?

File: ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs [hunk 7]
```diff
@@ -831,9 +832,58 @@ prop_TxSeq_lookupByTicketNo_sound smalls small =
...
+prop_TxSeq_splitAfterTxSizeSpec xs =
+    conjoin [ splitAndCheck txsz
+            | TxTicket _ _ txsz <- tickets ]
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs]: Wait a second, are you using the size of each transaction to split on? Then you won't get many interesting splits.

I suggest the following: introduce a data type `TxSizesAndMaxSize` (or something better) which consists of a `[Positive TxSizeInBytes]` and a `TxSizeInBytes`. The latter is generated using `choose (1, txSizes ..)`, so you'll get interesting spits. Let `prop_TxSeq_splitAfterTxSizeSpec` (and the other) take it as its argument. You can add a function from `TxSizesAndMaxSize` to `TxSeq Int`, like your local `txseq`, which you can reuse in both tests.

File: ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs [hunk 8]
```diff
@@ -831,9 +832,58 @@ prop_TxSeq_lookupByTicketNo_sound smalls small =
...
+
+    tickets :: [TxTicket Int]
+    tickets = txTickets txseq
+
+    splitAndCheck :: TxSizeInBytes -> Bool
+    splitAndCheck txsz =
+      let (implBefore, implAfter) = splitAfterTxSize txseq txsz
+          (specBefore, specAfter) = splitAfterTxSizeSpec txseq txsz
+      in    (txTickets implBefore) == (txTickets specBefore)
+         && (txTickets implAfter)  == (txTickets specAfter)
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs]: I would lose the parens here

File: ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs (hunk 6 above)
Comment (intricate) [reply ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs]: When I used `[Positive TxSizeInBytes]` I would run into a few counterexamples in which overflows were occurring when summing them up.

Comment (mrBliss) [reply ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs]: Then I would use `choose (1, x)` where `x` is some reasonable value. See my next comment about `TxSizesAndMaxSize`.

File: ouroboros-consensus/src/Ouroboros/Consensus/Mempool/TxSeq.hs (hunk 3 above)
Comment (intricate) [reply ouroboros-consensus/src/Ouroboros/Consensus/Mempool/TxSeq.hs]: Thanks, I didn't even know you could do that.

File: ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs (hunk 7 above)
Comment (intricate) [reply ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs]: Makes sense. I defined a type `TxSizeSplitTestSetup` (name is meh) that consists of a `[TxSizeInBytes]` (all of the generated tx sizes) and a `TxSizeInBytes` which is the actual size to split on in a test. This size to split on is generated using `choose (0, sumOfGeneratedTxSizes)`.

File: ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs [hunk 9]
```diff
@@ -831,9 +832,80 @@ prop_TxSeq_lookupByTicketNo_sound smalls small =
...
+  { tssTxSizes         :: ![TxSizeInBytes]
+  , tssTxSizeToSplitOn :: !TxSizeInBytes
+  } deriving Show
+
+instance Arbitrary TxSizeSplitTestSetup where
+  arbitrary = do
+    let txSizeMaxBound = 10000000 -- 10MB transaction max bound
+    txSizes <- listOf $ choose (1, txSizeMaxBound)
+    let totalTxSize = sum txSizes
+    txSizeToSplitOn <- choose (0, totalTxSize)
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs]: Just to make sure we're also testing the boundaries:
```haskell
frequency 
  [ (1, pure 0)
  , (8, choose (0, totalTxsSize))
  , (1, pure totalTxsSize)
  ]
```
(I would also rename `totalTxSize` to `totalTxsSize`)
There is some overlap now, but `choose (1, totalTxsSize - 1)` could underflow if `totalTxsSize == 0`, but you can change the generator if you prefer.

We should also test with a `txsSizeToSplitOn > totalTxsSize`:
```haskell
  , (1, choose (totalTxsSize + 1, totalTxsSize + 1000))
```

File: ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs [hunk 10]
```diff
@@ -831,9 +832,80 @@ prop_TxSeq_lookupByTicketNo_sound smalls small =
...
+      , tssTxSizeToSplitOn = txSizeToSplitOn
+      }
+
+  shrink TxSizeSplitTestSetup { tssTxSizes, tssTxSizeToSplitOn } =
+    [ TxSizeSplitTestSetup
+        { tssTxSizes         = tssTxSizes'
+        , tssTxSizeToSplitOn = tssTxSizeToSplitOn'
+        }
+    | tssTxSizes' <- shrinkList (const []) tssTxSizes
+    , tssTxSizeToSplitOn' <- shrinkIntegral tssTxSizeToSplitOn
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs]: This isn't the most efficient shrinker: what if we found a failure for `sum tssTxSize == tssTxSizeToSplitOn`? Now shrinking will probably generate tons of smaller non-relevant test cases. However, I don't think it's worth improving this (until we get actual test failures that are too large). So you can leave this as is.

File: ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs [hunk 11]
```diff
@@ -831,9 +832,80 @@ prop_TxSeq_lookupByTicketNo_sound smalls small =
...
+    ]
+
+-- | Convert a 'TxSizeSplitTestSetup' to a 'TxSeq'.
+txSizeSplitTestSetupToTxSeq :: TxSizeSplitTestSetup -> TxSeq Int
+txSizeSplitTestSetupToTxSeq TxSizeSplitTestSetup { tssTxSizes } =
+  TxSeq.toTxSeq $ zip3 (repeat 0) (repeat (TicketNo 0)) tssTxSizes
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs]: `[(0, TicketNo 0, tssTxSize) | tssTxSize <- tssTxSizes]`?

File: ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs [hunk 12]
```diff
@@ -831,9 +832,80 @@ prop_TxSeq_lookupByTicketNo_sound smalls small =
...
+
+    txseq :: TxSeq Int
+    txseq = txSizeSplitTestSetupToTxSeq tss
+
+    splitAndCheck :: TxSizeInBytes -> Property
+    splitAndCheck txsz =
+      let (implBefore, implAfter) = splitAfterTxSize txseq txsz
+          (specBefore, specAfter) = splitAfterTxSizeSpec txseq txsz
+      in      txTickets implBefore === txTickets specBefore
+         .&&. txTickets implAfter  === txTickets specAfter
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs]: I would simply inline this and add the `let`-bindings to the `where`-block.

File: ouroboros-consensus/src/Ouroboros/Consensus/Mempool/TxSeq.hs [hunk 13]
```diff
@@ -160,11 +180,57 @@ splitAfterTicketNo (TxSeq txs) n =
...
+
+    go :: TxSizeInBytes
+       -> [TxTicket tx]
+       -> [TxTicket tx]
+       -> ([TxTicket tx], [TxTicket tx])
+    go _ accTickets [] = (reverse accTickets, [])
+    go accByteSize accTickets (t:ts)
+      | let accByteSize' = accByteSize + txTicketTxSizeInBytes t
+      , accByteSize' <= n
+      = go (accByteSize + txTicketTxSizeInBytes t) (t:accTickets) ts
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Mempool/TxSeq.hs]:
```suggestion
      = go accByteSize' (t:accTickets) ts
```
Use it here too.

File: ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs [hunk 14]
```diff
@@ -830,9 +831,66 @@ prop_TxSeq_lookupByTicketNo_sound smalls small =
...
+
+data TxSizeSplitTestSetup = TxSizeSplitTestSetup
+  { tssTxSizes         :: ![TxSizeInBytes]
+  , tssTxSizeToSplitOn :: !TxSizeInBytes
+  } deriving Show
+
+instance Arbitrary TxSizeSplitTestSetup where
+  arbitrary = do
+    let txSizeMaxBound = 10 * 1024 * 1024 -- 10MB transaction max bound
+    txSizes <- listOf $ choose (1, txSizeMaxBound)
```
Comment (intricate) [inline ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs]: Oops. Meant to use `listOf1` after making the change specified here: https://github.com/input-output-hk/ouroboros-network/pull/1346#discussion_r358171400

File: ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs (hunk 9 above)
Comment (intricate) [reply ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs]: For the generation of `txSizes`, I've instead utilized `listOf1` to avoid the possibility of `sum txSizes == 0`.

Comment (mrBliss) [reply ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs]: No, `listOf` is fine, we want to correctly handle that case too.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs (hunk 2 above)
Comment (dcoutts) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node.hs]: The name is fine. The max block size actually comes from the current ledger protocol params (with the initial value in the genesis).

The block size refers to the total block size, including the block header and all other encoding overheads.

I suggest that we approximate this by a per-ledger-instance constant max overhead, so we would end up taking `maxBlockSize - blockEncodingOverheadsSize` worth of txs.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs [hunk 15]
```diff
@@ -257,6 +257,7 @@ mkNodeArgs registry cfg initState tracers btime chainDB = NodeArgs
     , blockFetchSize      = nodeBlockFetchSize
     , blockMatchesHeader  = nodeBlockMatchesHeader
     , maxUnackTxs         = 100 -- TODO
+    , maxBlockBodySize    = 2000000 -- TODO
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs]:
```suggestion
    , maxBlockBodySize    = 2_000_000 -- TODO
```
Needs the `NumDecimals` extension.

**Comment by dcoutts** [comment]
See also related #1354

File: ouroboros-consensus/src/Ouroboros/Consensus/Mempool/TxSeq.hs [hunk 16]
```diff
@@ -160,11 +180,57 @@ splitAfterTicketNo (TxSeq txs) n =
...
+  where
+    mapTuple :: (a -> b) -> (a, a) -> (b, b)
+    mapTuple f (x, y) = (f x, f y)
+
+    go :: TxSizeInBytes
+       -> [TxTicket tx]
+       -> [TxTicket tx]
+       -> ([TxTicket tx], [TxTicket tx])
+    go _ accTickets [] = (reverse accTickets, [])
+    go accByteSize accTickets (t:ts)
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Mempool/TxSeq.hs]: I would align on `accTickets`.

Personal preference:
```haskell
go accByteSize accTickets = \case
  []   -> ..
  t:ts -> ..

Comment (intricate) [reply ouroboros-consensus/src/Ouroboros/Consensus/Mempool/TxSeq.hs]: Switched to a lambda `case` with guards.

**Comment by mrBliss** [comment]
Great :+1:
